xla/service/cpu/onednn_contraction_rewriter.cc

```diff
@@ -586,148 +599,154 @@ class OneDnnContractionRewriteVisitor : public DfsHloRewriteVisitor {
-            ->mutable_optimization_config()
-            ->set_bias_broadcast(true);
-      }
-      TF_RETURN_IF_ERROR(matmul_call->set_backend_config(*backend_config));
+    // TODO(intel-tf): Remove this restriction once oneDNN has an optimized
+    // implementation for broadcasted add across all dimensions.
+    OneDnnFusionConfig_FusionKind kind = OneDnnFusionConfig::UNDEFINED;
+    kind =
+        (addend->shape().rank() == 1)
+            ? (GetKernelConfig<config>(&backend_config)->fusions().ops().empty()
+                   ? OneDnnFusionConfig::BIAS
+                   : OneDnnFusionConfig::UNDEFINED)
+            : OneDnnFusionConfig::BINARY_ADD;
+    if (kind == OneDnnFusionConfig::UNDEFINED) return absl::OkStatus();
 
-      HloInstruction* new_instr;
-      // If matched pattern has custom-call -> bitcast -> add, then we need to
-      // insert bitcast after the new fusion to maintain the correct shape
-      // (new-custom-call -> bitcast). Also, this will optionally be followed
-      // by -> convert for bf16 case to avoid datatype mismatch.
-      if (optional_dot_bitcast != nullptr &&
-          optional_dot_bitcast->opcode() == HloOpcode::kBitcast) {
-        if (optional_dot_convert != nullptr &&
-            optional_dot_convert->opcode() == HloOpcode::kConvert) {
-          auto bitcast_call =
-              matmul_call->AddInstruction(HloInstruction::CreateBitcast(
-                  ShapeUtil::ChangeElementType(
-                      instr->shape(), matmul_call->shape().element_type()),
-                  matmul_call));
-          new_instr =
-              bitcast_call->AddInstruction(HloInstruction::CreateConvert(
-                  ShapeUtil::ChangeElementType(
-                      bitcast_call->shape(),
-                      optional_dot_convert->shape().element_type()),
-                  bitcast_call));
-        } else {
-          new_instr = matmul_call->AddInstruction(
-              HloInstruction::CreateBitcast(instr->shape(), matmul_call));
-        }
+    GetKernelConfig<config>(&backend_config)->mutable_fusions()->add_ops(kind);
+
+    if (optional_addend_broadcast) {
+      GetKernelConfig<config>(&backend_config)
+          ->mutable_optimization_config()
+          ->set_bias_broadcast(true);
+    }
+    TF_RETURN_IF_ERROR(custom_call->set_backend_config(*backend_config));
+
+    HloInstruction* new_instr;
+    // If matched pattern has custom-call -> bitcast -> add, then we need to
+    // insert bitcast after the new fusion to maintain the correct shape
+    // (new-custom-call -> bitcast). Also, this will optionally be followed
+    // by -> convert for bf16 case to avoid datatype mismatch.
+    if (optional_contraction_bitcast != nullptr &&
+        optional_contraction_bitcast->opcode() == HloOpcode::kBitcast) {
+      if (optional_contraction_convert != nullptr &&
+          optional_contraction_convert->opcode() == HloOpcode::kConvert) {
+        auto bitcast_call =
```

Comment:
Given

```c++
 bitcast_call = custom_call->AddInstruction(HloInstruction::CreateBitcast(
                ShapeUtil::ChangeElementType(
                    instr->shape(), custom_call->shape().element_type()),
                custom_call));
```

As the result shape of the `bitcast_call->shape->element_type == custom_call->shape->element_type`, can we conclude that  the dimensions of `instr->shape` == dimensions of `custom_call->shape` ? If yes, isn't the `bitcast_call` created here is a no-op.  

I was expecting something like following here

```c++
auto bitcast_call = custom_call->AddInstruction(
            HloInstruction::CreateBitcast(instr->shape(), custom_call));
```

Reply:
The dimensions of ```instr->shape()``` might not always be equal to the dimensions of ```custom-call->shape()```. A hypothetical example:

``` 

op1               op2                                                           op1               op2                                         op1               op2              addend 
 |                 |                                                             |                 |                                           |                 |                 |
 |_________________|                                                             |_________________|                                           |_________________|_________________|
          |                                                                               |                                                                      |
          |                                                                               |                                                                      |
     custom-call                                                                     custom-call          addend                                            custom-call    
      [1024,100]                                                                      [1024,100]           [100]                                            [1024,100]
          |                                                                               |                 |                                                    |
          |                                                                               |_________________|                                                    |
       bitcast    (broadcasted) addend                    =======>                                 |                        =======>                          bitcast
     [32,32,100]       [32,32,100]                                                                 |                                                        [32,32,100]
          |                 |                                                                  add (instr)
          |_________________|                                                                  [1024,100]             
                   |                                                                               | 
                   |                                                                               |
               add (instr)                                                                      bitcast
               [32,32,100]                                                                    [32,32,100]

                                                                        Note: oneDNN custom-calls can absorb non-broadcasted addends
     
```


Also, what you suggested in the last line is exactly what we have for the non-BF16 case.
https://github.com/openxla/xla/blob/8d10441bcf6c2e23fb2e17e6321bc31e29f0772b/xla/service/cpu/onednn_contraction_rewriter.cc#L734-L735


 The BF16 case gets a little tricky because currently we don't support elemental arithmetic operations like Add in the lower precision, as a result they are upcated to F32. Hence the pattern looks something like this:
Here, we need to change the element type of the generated ```bitcast``` so that it matches the ```custom-call``` dtype.
``` 

op1               op2                                        op1               op2              addend 
 |                 |                                          |                 |                 |
 |_________________|                                          |_________________|_________________|
          |                                                                     |
          |                                                                     |
     custom-call                                                           custom-call    
   BF16 [1024,100]                        =======>                       BF16 [1024,100]
          |                                                                     |
          |                                                                     |
       convert                                                               bitcast
    F32 [1024,100]                                                       BF16 [32,32,100]
          |                                                  
          |
       bitcast    (broadcasted and upcasted) addend                                                
    F32 [32,32,100]    F32 [32,32,100]                                                        
          |                 |                                                                  
          |_________________|                                                                             
                   |                                                                            
                   |                                                                            
               add (instr)                                                                
             F32 [32,32,100]                                                                  
                   |
                   |                                                     
                convert
            BF16 [32,32,100] 
```

Please let me know if this helps answer your questions.